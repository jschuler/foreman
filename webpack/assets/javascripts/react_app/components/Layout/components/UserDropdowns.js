import React from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';
import { Icon } from 'patternfly-react';
import {
  Dropdown,
  DropdownToggle,
  DropdownItem,
  DropdownSeparator,
} from '@patternfly/react-core';

import { translate as __ } from '../../../common/I18n';

class UserDropdowns extends React.Component {
  constructor(props) {
    super(props);
    this.state = { userDropdownOpen: false };
    this.onDropdownToggle = userDropdownOpen => {
      this.setState({
        userDropdownOpen,
      });
    };
    this.onDropdownSelect = event => {
      this.setState({
        userDropdownOpen: !this.state.userDropdownOpen,
      });
    };
  }

  render() {
    const {
      activeKey, // eslint-disable-line react/prop-types
      activeHref, // eslint-disable-line react/prop-types
      user,
      changeActiveMenu,
      notificationUrl,
      stopImpersonationUrl,
      ...props
    } = this.props;

    const { userDropdownOpen } = this.state;

    const userInfo = get(user, 'current_user.user');

    const userDropdownItems = user.user_dropdown[0].children.map((item, i) =>
      item.type === 'divider' ? (
        <DropdownSeparator key={i} />
      ) : (
        <DropdownItem
          key={i}
          href={item.url}
          onClick={() => {
            changeActiveMenu({ title: 'User' });
          }}
          className="user_menuitem"
          {...item.html_options}
        >
          {__(item.name)}
        </DropdownItem>
      )
    );

    return (
      <React.Fragment>
        {userInfo && (
          <Dropdown
            isPlain
            position="right"
            onSelect={this.onDropdownSelect}
            isOpen={userDropdownOpen}
            toggle={
              <DropdownToggle onToggle={this.onDropdownToggle}>
                <Icon type="fa" name="user avatar small" />
                {userInfo.name}
              </DropdownToggle>
            }
            dropdownItems={userDropdownItems}
            {...props}
          />
        )}
      </React.Fragment>
    );
  }
}

UserDropdowns.propTypes = {
  /** Additional element css classes */
  className: PropTypes.string,
  /** User Data Array */
  user: PropTypes.object,
  /** notification URL */
  notificationUrl: PropTypes.string,
  /** changeActiveMenu Func */
  changeActiveMenu: PropTypes.func,
  stopImpersonationUrl: PropTypes.string,
};
UserDropdowns.defaultProps = {
  className: '',
  user: {},
  notificationUrl: '',
  changeActiveMenu: null,
  stopImpersonationUrl: '',
};
export default UserDropdowns;
