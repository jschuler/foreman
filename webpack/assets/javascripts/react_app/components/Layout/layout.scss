@import '../../common/colors.scss';

.pf-c-page.foreman-theme {
  .pf-c-page__header {
    .pf-c-data-toolbar {
      background-color: transparent;
    }
  }

  .pf-c-page__sidebar.pf-m-dark {
    background-color: $nav-pf-vertical-secondary-active-bg-color;

    .pf-c-nav.pf-m-dark > .pf-c-nav__list {
      .pf-c-nav__item.pf-m-current,
      .pf-c-nav__item > .pf-c-nav__link.pf-m-current {
        background-color: $nav-pf-vertical-active-bg-color;
      }

      .pf-c-nav__item.pf-m-current > .pf-c-nav__link[data-component=pf-nav-expandable] {
        color: $nav-pf-vertical-active-icon-color;
      }

      .pf-c-nav__link {
        color: $nav-pf-vertical-icon-color;
      }

      .pf-c-nav__simple-list .pf-c-nav__link {
        &.pf-m-current,
        &.pf-m-hover,
        &:hover {
          background-color: $nav-pf-vertical-active-bg-color;
          color: $nav-pf-vertical-active-icon-color;
        }
      }
    }
  }
}

#account_menu {
  outline: none;
}

#vertical-spinner {
  margin-top: 15px;
  margin-left: 5px;
}
