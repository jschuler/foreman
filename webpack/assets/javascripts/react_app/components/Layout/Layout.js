import React from 'react';
import {
  Brand,
  Page,
  PageSection,
  PageSectionVariants,
  PageSidebar,
  SkipToContent,
} from '@patternfly/react-core';
// eslint-disable-next-line
import globalBreakpointMd from '@patternfly/react-tokens/dist/js/global_breakpoint_md';
import CustomPageHeader from './components/CustomPageHeader';
import {
  getActive,
  getCurrentPath,
  layoutPropTypes,
  layoutDefaultProps,
} from './LayoutHelper';
import VerticalNav from './components/VerticalNav';
import Toolbar from './components/Toolbar';
import './layout.scss';

class Layout extends React.Component {
  componentDidMount() {
    const {
      items,
      data,
      fetchMenuItems,
      changeLocation,
      currentLocation,
      changeOrganization,
      currentOrganization,
      changeActiveMenu,
      activeMenu,
    } = this.props;
    if (items.length === 0) fetchMenuItems(data);

    const activeURLMenu = getActive(data.menu, getCurrentPath());
    if (activeMenu !== activeURLMenu.title) {
      changeActiveMenu(activeURLMenu);
    }

    if (
      data.taxonomies.locations &&
      !!data.locations.current_location &&
      currentLocation !== data.locations.current_location
    ) {
      const initialLocTitle = data.locations.current_location;
      const initialLocId = data.locations.available_locations.find(
        loc => loc.title === initialLocTitle
      ).id;
      changeLocation({ title: initialLocTitle, id: initialLocId });
    }

    if (
      data.taxonomies.organizations &&
      !!data.orgs.current_org &&
      currentOrganization !== data.orgs.current_org
    ) {
      const initialOrgTitle = data.orgs.current_org;
      const initialOrgId = data.orgs.available_organizations.find(
        org => org.title === initialOrgTitle
      ).id;
      changeOrganization({ title: initialOrgTitle, id: initialOrgId });
    }
  }

  render() {
    const {
      items,
      data,
      isLoading,
      changeActiveMenu,
      changeOrganization,
      changeLocation,
      currentOrganization,
      currentLocation,
      activeMenu,
      children,
      history,
    } = this.props;

    const afterNavToggle = navOpen => {
      const mobileView =
        window.innerWidth < Number.parseInt(globalBreakpointMd.value, 10);
      if (mobileView) {
        // ignore for mobile view
        return;
      }
      // toggles a class in the body tag, so that the main #rails-app-content container can have the appropriate width
      if (navOpen) {
        document.body.classList.add('pf-m-expanded');
      } else {
        document.body.classList.remove('pf-m-expanded');
      }
    };
    const header = (
      <CustomPageHeader
        logo={
          <Brand
            src={data.logo}
            alt={data.brand}
            style={{ marginTop: '8px' }}
          />
        }
        logoProps={{ href: data.root }}
        showNavToggle
        contextSelector={
          <Toolbar
            data={data}
            currentLocation={currentLocation}
            changeLocation={changeLocation}
            currentOrganization={currentOrganization}
            changeOrganization={changeOrganization}
            isLoading={isLoading}
            changeActiveMenu={changeActiveMenu}
          />
        }
        afterNavToggle={afterNavToggle}
        className="navbar-pf-vertical"
      />
    );
    const sidebar = (
      <PageSidebar
        nav={
          <VerticalNav
            items={items}
            activeMenu={activeMenu}
            changeActiveMenu={changeActiveMenu}
            history={history}
          />
        }
        theme="dark"
      />
    );
    const pageId = 'main';
    const pageSkipToContent = (
      <SkipToContent href={`#${pageId}`}>Skip to content</SkipToContent>
    );

    return (
      <Page
        header={header}
        sidebar={sidebar}
        isManagedSidebar
        skipToContent={pageSkipToContent}
        mainContainerId={pageId}
        className="foreman-theme"
      >
        <PageSection variant={PageSectionVariants.light}>
          {children}
        </PageSection>
      </Page>
    );
  }
}

Layout.propTypes = {
  ...layoutPropTypes,
};

Layout.defaultProps = {
  ...layoutDefaultProps,
};

export default Layout;
