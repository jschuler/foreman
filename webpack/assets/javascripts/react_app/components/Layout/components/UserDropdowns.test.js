import React from 'react';
import { mount } from '@theforeman/test';
import { testComponentSnapshotsWithFixtures } from '../../../common/testHelpers';

import UserDropdowns from '../components/UserDropdowns';
import { userDropdownProps } from '../Layout.fixtures';

const createStubs = () => ({
  changeActiveMenu: jest.fn(),
  isOpen: true,
});

const fixtures = {
  'render switcher w/loading': {
    ...userDropdownProps,
    ...createStubs(),
  },
};

describe('UserDropdown', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(UserDropdowns, fixtures));

  describe('trigger onClicks', () => {
    const wrapper = mount(
      <UserDropdowns {...userDropdownProps} {...createStubs()} />
    );
    wrapper
      .find('.user_menuitem')
      .last()
      .simulate('click');
  });
});
