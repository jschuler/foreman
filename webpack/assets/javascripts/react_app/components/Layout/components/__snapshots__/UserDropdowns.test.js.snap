// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDropdown rendering render switcher w/loading 1`] = `
<Fragment>
  <Dropdown
    className=""
    dropdownItems={
      Array [
        <DropdownItem
          className="user_menuitem"
          href="/"
          onClick={[Function]}
        >
          My Account
        </DropdownItem>,
        <DropdownSeparator />,
      ]
    }
    isOpen={true}
    isPlain={true}
    notification_url="/"
    onSelect={[Function]}
    position="right"
    toggle={
      <DropdownToggle
        onToggle={[Function]}
      >
        <Icon
          name="user avatar small"
          type="fa"
        />
        G L
      </DropdownToggle>
    }
  />
</Fragment>
`;
